### PR TITLE
Remove Implicit rank promotion

### LINKED ISSUE #1534: Avoid rank promotion in Flax codebase
Silent rank promotion can conceal errors that would otherwise throw exception, so I always prefer to run my code with
```
from jax.config import config
config.update("jax_numpy_rank_promotion", "raise")
```
This way an exception is thrown if rank promotion happens anywhere (https://jax.readthedocs.io/en/latest/rank_promotion_warning.html).

Unfortunately this appears incompatible with flax, since it uses rank promotion in biases, normalization, some examples etc, so I can't use the above setting to catch unintentional rank promotion in my code.

It would be very nice if flax could be rank promotion-free. Thanks a lot for the great library!

### PR DESCRIPTION
**Pull request by jheek**
- users can now disable implicit rank promotion in JAX without Flax throwing errors.
- Tests are no run with JAX_NUMPY_RANK_PROMOTION=raise
- This change uncovered 2 faulty LSTM tests.

Fixes #1534

### CODE REVIEW

File: tests/run_all_tests.sh [hunk 1]
```diff
@@ -42,7 +43,7 @@ handle_errors () {
 sphinx-build -M doctest docs docs/_build
 
 # Run battery of core FLAX API tests.
-pytest -n 4 tests $PYTEST_OPTS
+pytest -n auto tests $PYTEST_OPTS
```
Comment (levskaya) [inline tests/run_all_tests.sh:46]: didn't know about `auto`

**Review by levskaya — APPROVED** [review]
This all looks great!  Thanks so much for isolating all of these.

File: tests/run_all_tests.sh (hunk 1 above)
Comment (jheek) [reply tests/run_all_tests.sh:46]: We had it before, It was accidentally removed in the linen refactor
